cli/jsApiSpec.ts

```diff
@@ -0,0 +1,133 @@
+        case 'string':
+            return 'string'
+        case 'bytes':
+            return 'ArrayBuffer'
+        case 'any':
+            return 'any'
+        case 'literal':
+            if (typeof argsJson.value === 'string') {
+                return `"${argsJson.value}"` as string
+            } else {
+                return argsJson.value!.toString()
+            }
+        case 'id':
+            return `Id<"${argsJson.tableName}">`
+        case 'array':
+            return `Array<${generateArgsType(argsJson.value)}>`
+        case 'record':
+            return 'any'
+        case 'object': {
+            const members: string[] = Object.entries(argsJson.value).map(
+                ([key, value]) => {
+                    return `${key}${value.optional ? '?' : ''}: ${generateArgsType(
+                        value.fieldType
+                    )},`
+                }
+            )
+            if (members.length === 0) {
+                // special case empty object
+                return 'Record<string, never>'
+            }
+            return `{ ${members.join('\n')} }`
+        }
+        case 'union': {
+            const members: string[] = argsJson.value.map((v) => generateArgsType(v))
+            return members.join(' | ')
+        }
+    }
+}
+
+function generateApiType(tree: Record<string, any>) {
+    const isFunction = tree.functionType !== undefined;
+    if (isFunction) {
+        const output =
+            tree.returns === null || tree.returns === undefined
+                ? 'any'
+                : generateArgsType(tree.returns);
+        return `FunctionReference<"${(tree.functionType as string).toLowerCase()}", "${tree.visibility.kind
+            }", ${generateArgsType(tree.args)}, ${output}>`
+    }
+    const members: string[] = Object.entries(tree).map(([key, value]) => {
+        return `${key}: ${generateApiType(value)}`
+    })
+    return `{ ${members.join('\n')} }`
+}
+
+export function generateApiSpec(functionSpec: FunctionSpec) {
+    const publicFunctionTree: Record<string, any> = {};
+    const internalFunctionTree: Record<string, any> = {};
+    for (const fn of functionSpec.functions) {
+        const [modulePath, functionName] = fn.identifier.split(':');
+        const withoutExtension = modulePath.slice(0, modulePath.length - 3);
+        const pathParts = withoutExtension.split('/');
+        let treeNode =
+            fn.visibility.kind === "internal"
+                ? internalFunctionTree
+                : publicFunctionTree;
+        for (let i = 0; i < pathParts.length; i += 1) {
+            const pathPart = pathParts[i];
+            if (treeNode[pathPart] === undefined) {
+                treeNode[pathPart] = {};
+            }
+            treeNode = treeNode[pathPart];
+        }
+        treeNode[functionName] = fn;
+    }
+    const apiType = generateApiType(publicFunctionTree);
+    const internalApiType = generateApiType(internalFunctionTree);
+    return (`
```

Comment:
Could we run prettier first? I know we do this for codegen from the main convex CLI, so perhaps we can pattern match.